src/gc.c

```diff
@@ -2805,19 +2807,46 @@ JL_DLLEXPORT void *jl_realloc(void *p, size_t sz)
     return (void *)(pnew + 2);
 }
 
-JL_DLLEXPORT void *jl_gc_managed_malloc(size_t sz)
+// TODO add special casing for macOS, where there is guarantee of alignment
+// TODO add special casing for 64 bit systems when 16 byte alignment is requested
+// TODO add checks on align?
+void *jl_gc_malloc_aligned(size_t sz, size_t align)
 {
-    jl_ptls_t ptls = jl_get_ptls_states();
-    maybe_collect(ptls);
-    size_t allocsz = LLT_ALIGN(sz, JL_CACHE_BYTE_ALIGNMENT);
-    if (allocsz < sz)  // overflow in adding offs, size was "negative"
-        jl_throw(jl_memory_exception);
-    gc_num.allocd += allocsz;
-    gc_num.malloc++;
-    void *b = malloc_cache_align(allocsz);
-    if (b == NULL)
-        jl_throw(jl_memory_exception);
-    return b;
+    void *p0 = jl_gc_malloc(sz + align);
```

Comment:
Come to think of it, is it really correct for `jl_gc_counted_malloc` to add `JL_SMALL_BYTE_ALIGNMENT` to the size? It doesn't return aligned pointers itself, so it seems to me the caller should have to handle the size adjustment.